In [1]:
import pandas as pd

In [168]:
userComb = pd.read_csv("user_info_comb.csv")

/Users/yimengz/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (23) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


127476 users

In [169]:
userComb.drop(['password','id.1','primary_address_id','lat', 'lng'], axis=1, inplace=True)
userComb.head()

last_login is_superuser  id    username  \
0  2017-01-27 17:42:31.857178+00            t   1  migueldv90   
1  2014-12-31 03:52:34.318484+00            f   2        Zhuf   
2  2014-12-31 15:14:49.786013+00            f   3   oxheywood   
3         2015-05-24 01:09:33+00            f   4    mimobrow   
4  2017-01-07 15:51:51.854495+00            f   5   joetroyen   

                      email first_name last_name gender  tier  \
0  miguel@penpalschools.com     Miguel   Vazquez      M     4   
1            zhuf@sfusd.edu  Fengsheng       Zhu    NaN     2   
2    mikki.caplan@gmail.com      Mikki    Caplan    NaN     2   
3                       NaN    Michael    Monroe    NaN     1   
4     joe@penpalschools.com        Joe    Troyen    NaN     4   

                                                 img  \
0  user_app/1/08_07_2015_18_26_46_736474_upload-i...   
1                                default-profile.svg   
2                                default-profile.svg   
3                                default-profile.svg   
4  user_app/5/04_01_2015_17_59_50_754213_upload-i...   

              ...              email_verified points  \
0             ...                           f      0   
1             ...                           f      0   
2             ...                           f      0   
3             ...                           f      0   
4             ...                           f      0   

                    img_activity street_line1 street_line2 zipcode     city  \
0  2016-09-09 04:17:40.075083+00          NaN          NaN     NaN   Austin   
1  2016-09-09 04:17:40.075083+00          NaN          NaN     NaN       SF   
2  2016-09-09 04:17:40.075083+00          NaN          NaN     NaN  Watford   
3  2016-09-09 04:17:40.075083+00          NaN          NaN     NaN   Austin   
4  2016-09-09 04:17:40.075083+00          NaN          NaN     NaN   Austin   

      state          country            formatted_address  
0        TX    United States              Austin, TX, USA  
1       CA    United States       San Francisco, CA, USA   
2  England   United Kingdom   Watford, Hertfordshire, UK   
3       TX    United States              Austin, TX, USA   
4       TX    United States              Austin, TX, USA   

[5 rows x 26 columns]

In [170]:
userComb.columns.values

array(['last_login', 'is_superuser', 'id', 'username', 'email',
       'first_name', 'last_name', 'gender', 'tier', 'img', 'bio',
       'stripe_id', 'is_staff', 'is_active', 'date_joined', 'dob',
       'email_verified', 'points', 'img_activity', 'street_line1',
       'street_line2', 'zipcode', 'city', 'state', 'country',
       'formatted_address'], dtype=object)

In [167]:
userComb['primary_address_id'].count()

127329

In [44]:
userComb.describe(include=['O'])

last_login is_superuser username  \
count                          127476       127476   127476   
unique                         127470            2   127476   
top     2015-10-07 15:09:50.107519+00            f  KateNic   
freq                                2       127474        1   

                        email first_name last_name gender  \
count                   52645      97654     95666  93454   
unique                  50732      27800      3446      2   
top     porurpenpal@gmail.com      Emily         M      F   
freq                       81        457      7363  48932   

                        img      bio           stripe_id        ...         \
count                127472    15649                  57        ...          
unique                23029    15153                  57        ...          
top     default-profile.svg  Teacher  cus_9rpzLyBlQNoNUA        ...          
freq                  75451       73                   1        ...          

                                  dob email_verified  \
count                          127476         127476   
unique                          95895              2   
top     2015-03-20 03:20:34.472988+00              f   
freq                             9243         122412   

                         img_activity street_line1 street_line2 zipcode  \
count                          127476          585            5    3971   
unique                          39788          433            1    2355   
top     2016-09-09 04:17:40.075083+00     New York                48118   
freq                            87652           21            5      62   

          city  state country formatted_address  
count    24930  25637  127468             27112  
unique    6691   1846     308             10189  
top     Austin     TX    None   Austin, TX, USA  
freq       482   1195  100363               347  

[4 rows x 23 columns]

### last_login

In [112]:
print(type(userComb["last_login"][0]))

<class 'str'>


### is_superuser

In [141]:
userComb.is_superuser=userComb.is_superuser.map({'t':0,'f':1})
userComb.groupby('is_superuser')['id','username'].count()

id  username
is_superuser                  
0                  2         2
1             127474    127474

<p> Because there is only 2 super-user, they may be outliers. Let us see who are the two super-user</p>

In [130]:
userComb[userComb.is_superuser==0]

last_login  is_superuser  id      username  \
0   2017-01-27 17:42:31.857178+00             0   1    migueldv90   
24  2017-01-26 20:58:10.536884+00             0  32  markdanforth   

                          email first_name last_name gender  tier  \
0      miguel@penpalschools.com     Miguel   Vazquez      M     4   
24  mark.danforth@bookaround.co     PenPal     Coach    NaN     4   

                                                  img    ...      \
0   user_app/1/08_07_2015_18_26_46_736474_upload-i...    ...       
24                                default-profile.svg    ...       

                     img_activity  street_line1 street_line2 zipcode    city  \
0   2016-09-09 04:17:40.075083+00           NaN          NaN     NaN  Austin   
24         2016-09-09 04:17:40+00           NaN          NaN     NaN  Austin   

   state        country formatted_address        lat        lng  
0     TX  United States   Austin, TX, USA  30.267153 -97.743061  
24    TX  United States   Austin, TX, USA  30.267153 -97.743061  

[2 rows x 29 columns]

They are migueldv and markdanforth, who are the employees from Penpal Schools. They may just test the website functions. So I decide to delete the is_superuser colunm.

In [142]:
userComb.drop(['is_superuser'], axis=1, inplace=True)

### username
<p> Every user name is unique. Nothing can be done with it.

### email
<p> It is wired that there are 52645 user have input their email but only 50732 unique values are found. And porurpenpal@gmail.com appears 81 times. Whoes email is porurpenpal@gmail.com?

In [124]:
userComb[userComb.email=="porurpenpal@gmail.com"]

last_login     id         username  \
54598  2016-01-21 09:18:10.746063+00  56997     ajaydevanand   
54651  2016-02-03 11:25:30.152268+00  57050        annamalai   
54671  2016-01-21 09:54:45.760968+00  57070     deepakprasad   
54683  2016-01-21 09:57:47.744758+00  57082             deva   
54703  2016-01-21 10:02:34.512706+00  57102         dhanushc   
54712  2016-01-21 10:05:59.040857+00  57111         dhanushm   
54726  2016-01-21 10:12:04.757306+00  57125         elamaran   
54730  2016-01-21 10:15:01.495477+00  57129       gunasekark   
54741  2016-01-21 10:22:45.708912+00  57140      harishporur   
54747  2016-01-21 10:25:24.854553+00  57146          harishv   
54750  2016-06-13 11:22:26.138069+00  57149          hemanth   
54761   2016-01-21 10:33:17.62967+00  57160        karthickc   
54769  2016-01-21 10:35:44.151384+00  57168     kishoreporur   
54773  2016-01-21 10:38:07.421344+00  57172           logesh   
54777  2016-01-21 10:41:01.597068+00  57176     logeshwarans   
54781  2016-01-21 10:44:15.631465+00  57180  manikandanporur   
54783  2016-02-05 06:04:58.655367+00  57182      manikandanv   
54795  2016-03-28 10:06:48.321467+00  57194           ramesh   
54796  2016-01-21 10:55:16.269132+00  57195           mukesh   
54800  2016-01-21 11:00:15.898681+00  57199          naveenb   
54803  2016-01-21 11:02:58.888293+00  57202      naveenporur   
54805  2016-01-21 11:05:45.429763+00  57204         pradeepr   
54806  2016-01-21 11:07:49.725723+00  57205    pradeepkumarn   
54809  2016-01-21 11:10:00.728165+00  57208      sakthikumar   
54812  2016-01-21 11:13:02.722124+00  57211       samsamuvel   
54814   2016-01-21 11:13:51.73347+00  57213          abijohn   
54815  2016-01-21 11:15:41.130021+00  57214          sanjayv   
54818  2016-01-21 11:17:27.247416+00  57217     sarveshwaran   
54820  2016-01-21 11:20:11.071385+00  57219          sathyas   
54822  2016-01-21 11:23:15.248743+00  57221        shanmugam   
...                              ...    ...              ...   
54886  2016-01-21 12:17:02.638071+00  57286          jaiaksh   
54887  2016-02-04 11:58:33.092175+00  57287         dhineshn   
54890  2016-01-21 12:20:30.726115+00  57290         dhineshp   
54891   2016-01-21 12:22:07.07324+00  57291           joshep   
54892  2016-01-21 12:23:56.990423+00  57292         dhineshr   
54893  2016-01-21 12:25:24.560749+00  57293        karthickm   
54896  2016-01-21 12:28:16.244108+00  57296           gokult   
54897  2016-01-21 12:30:28.766085+00  57297   hariharanporur   
54898  2016-01-21 12:30:47.083328+00  57298        kathirvel   
54899  2016-01-21 12:33:14.227525+00  57299           kananc   
54901  2016-01-21 12:34:44.208998+00  57301    kishorekumarg   
54902  2016-01-21 12:35:34.108901+00  57302     karthikporur   
54904  2016-01-21 12:37:37.260861+00  57304          lokeshs   
54909  2016-01-21 12:40:46.229263+00  57309          lokeshr   
54911  2016-01-21 12:42:20.567559+00  57311    muralidharanm   
54915  2016-01-21 12:48:05.003773+00  57316          lokeshv   
59755  2016-02-03 11:10:36.702872+00  62208    dhanushcporur   
60061  2016-02-04 10:27:59.932672+00  62525    dhanushporur.   
60108  2016-06-06 10:50:23.969024+00  62572       abiabraham   
60110  2016-02-04 12:24:39.130251+00  62574      prakashraj1   
60112  2016-02-04 12:29:00.776996+00  62576      sanjayporur   
60113  2016-02-04 12:26:41.027474+00  62577        santhoshk   
60115  2016-02-04 12:29:31.409835+00  62579        saravanak   
60118   2016-02-04 12:32:36.38809+00  62582     sathishporur   
60120  2016-02-04 12:34:44.803026+00  62584     sathishkumar   
61496  2016-06-13 12:04:16.056664+00  63979     Sharpchandru   
61601  2016-02-23 09:08:13.838738+00  64086          balajik   
61602  2016-02-29 12:34:18.914369+00  64087       aslinpraba   
61603  2016-02-11 08:54:35.542104+00  64088   karthikeyan001   
62087  2016-02-15 12:01:53.256798+00  64579            raghu   

                       email    first_name last_name  gen

All the last names are one single letter, same gender, same tier, most of them use deful image, and last login are around Jan/Feb 2016. So those data is fake data, it should be deleted.

In [143]:
userComb=userComb[userComb.email!="porurpenpal@gmail.com"]

In [161]:
userComb.groupby('email')['id'].count().reset_index(name='count').sort_values(['count'], ascending=False).head(25)

email  count
39626         penpalmugappair@gmail.com     53
23410          hemapenpal2016@gmail.com     41
21546    garland.a@rochesterschools.com     39
39627          penpalmugappir@gmail.com     34
33562             manish.sharma@aif.org     27
43808         sarah.hodgkinson@risd.org     26
39592                pelusocl@yahoo.com     26
23744             holycrossy6@gmail.com     25
27736                 jscott@cpsk12.org     24
46183  stevenson.d@rochesterschools.com     23
23773               hopectckn@gmail.com     23
24869           j2.huntersbar@gmail.com     22
24380    info@gangnamenglishacademy.com     21
48584           ushasurendars@gmail.com     21
38078               nidhi.goyal@aif.org     18
26146            jenny.kelly@dsisdtx.us     17
15967          computerlabwms@gmail.com     16
25430                 jaredj@jsd171.org     16
49527       wigleym@highlands.k12.fl.us     15
21812           gghsmullana90@gmail.com     14
27161          johnsonsclass3@yahoo.com     13
16345      crockerfarmroom204@gmail.com     13
42167                rs761090@gmail.com     13
35388        metropolitan2006@gmail.com     13
28811       katrina@communityschool.net     11

<p> Same thing happens with "abhi.gladiator210@gmail.com" and "penpalmugappair@gmail.com"<p>

In [162]:
userComb.describe(include=['O'])

last_login username                      email  \
count                         127331   127331                      52500   
unique                        127325   127331                      50730   
top     2015-10-13 15:12:18.20282+00  KateNic  penpalmugappair@gmail.com   
freq                               2        1                         53   

       first_name last_name gender                  img      bio  \
count       97509     95521  93310               127327    15647   
unique      27745      3446      2                23017    15151   
top         Emily         M      F  default-profile.svg  Teacher   
freq          457      7360  48893                75318       73   

                 stripe_id is_staff        ...         \
count                   57   127331        ...          
unique                  57        2        ...          
top     cus_9rpzLyBlQNoNUA        f        ...          
freq                     1   127326        ...          

                                  dob email_verified  \
count                          127331         127331   
unique                          95857              2   
top     2015-03-20 03:20:34.472988+00              f   
freq                             9243         122267   

                         img_activity street_line1 street_line2 zipcode  \
count                          127331          585            5    3971   
unique                          39788          433            1    2355   
top     2016-09-09 04:17:40.075083+00     New York                48118   
freq                            87507           21            5      62   

          city  state country formatted_address  
count    24930  25637  127323             27112  
unique    6691   1846     308             10189  
top     Austin     TX    None   Austin, TX, USA  
freq       482   1195  100218               347  

[4 rows x 22 columns]

In [163]:
userComb[userComb.email=="hemapenpal2016@gmail.com"]

last_login     id         username  \
27705  2016-02-15 06:05:43.606963+00  29867        BHARATH.K   
27706  2016-02-06 09:05:48.648465+00  29868       GOKULRAJ.S   
27707  2016-02-08 08:43:44.034413+00  29869     MANIKANDAN.M   
27708  2016-02-06 09:08:11.104255+00  29870    NAVEENKUMAR.A   
27709  2016-02-06 09:09:17.974234+00  29871    PUGAZHENTHI.I   
27712  2016-02-06 09:10:02.333227+00  29874         SANJAY.M   
27715  2016-02-08 08:46:14.475709+00  29877  THAMEEMANSARI.K   
27717  2016-02-08 08:51:00.276792+00  29879        ARCHANA.I   
27719  2016-02-08 09:12:23.238637+00  29881    BELCIYAMARY.A   
27721  2016-02-06 09:18:45.362934+00  29883      GUNAVATHY.M   
27722  2016-02-15 06:31:11.738082+00  29884    JEEVABLESSI.T   
27723  2016-02-06 09:24:17.021658+00  29885   JENIFERANGEL.D   
27724  2016-02-08 09:45:41.407624+00  29886        LAVANYA.R   
27725  2016-02-06 09:27:34.612229+00  29887     MADHIMALAR.B   
27726  2016-02-15 06:32:53.540014+00  29888        NATHIYA.V   
27727  2016-02-06 09:38:15.724535+00  29889       SANDHIYA.D   
27728  2016-02-08 09:53:56.968163+00  29890       THAMARAI.S   
28000  2016-02-03 06:32:49.264062+00  30163   AROKIYASUGAN.C   
28002  2016-02-03 06:31:03.750866+00  30165         DEEPAK.I   
28003  2016-02-03 06:27:48.434141+00  30166        KISHORE.S   
28004  2016-02-03 06:24:50.458476+00  30167    KISHORERAJE.M   
28005  2016-02-03 06:19:25.011559+00  30168         LOKESH.S   
28006  2016-02-03 06:12:46.442735+00  30169        NAGILAN.P   
28007  2016-02-03 06:11:28.064887+00  30170         NAVEEN.B   
28009  2016-02-02 11:17:25.139412+00  30172        PRAKASH.K   
28010  2016-02-03 05:54:48.550844+00  30173       SANTHOSH.D   
28011  2016-02-03 05:57:07.363922+00  30174        VIGNESH.K   
28012  2016-02-03 06:02:03.046155+00  30175   VIGNESHWARAN.R   
28013  2016-02-03 06:04:14.677561+00  30176         YOGESH.E   
28014   2016-02-03 06:38:44.86967+00  30177        ABINAYA.S   
28015  2016-02-09 05:54:23.279865+00  30178          DIVYA.M   
28016  2016-02-03 06:45:22.293644+00  30179      KIRUTHIGA.S   
28017  2016-02-03 06:46:45.787657+00  30180        KUMUTHA.K   
28018  2016-02-03 06:53:34.329187+00  30181        NIKITHA.D   
28019  2016-02-03 06:56:24.957478+00  30182        PREETHI.B   
28020  2016-02-03 08:58:12.823275+00  30183          RAMYA.S   
28021  2016-02-03 08:59:22.505131+00  30184        SHAKANA.S   
28022  2016-02-18 13:19:54.095592+00  30185       SANDHIYA.K   
28023  2016-02-03 09:04:32.801242+00  30186       SHARMILA.A   
35989  2016-02-03 06:48:23.992604+00  38168      MAHESWARI.R   
35990  2016-02-03 09:12:47.566233+00  38169         SWETHA.K   

                          email     first_name last_name gender  tier  \
27705  hemapenpal2016@gmail.com        Bharath         K      M     1   
27706  hemapenpal2016@gmail.com       gokulraj         S      M     1   
27707  hemapenpal2016@gmail.com     Manikandan         M      F     1   
27708  hemapenpal2016@gmail.com   Naveen Kumar         A      M     1   
27709  hemapenpal2016@gmail.com    pugazhenthi         I      M     1   
27712  hemapenpal2016@gmail.com         sanjay         M      M     1   
27715  hemapenpal2016@gmail.com  thameemansari         K      M     1   
27717  hemapenpal2016@gmail.com        archana         I      F     1   
27719  hemapenpal2016@gmail.com    Belciyamary         A      F     1   
27721  hemapenpal2016@gmail.com      gunavathy         M      F     1   
27722  hemapenpal2016@gmail.com    jeevablessi         T      F     1   
27723  hemapenpal2016@gmail.com   jeniferangel         D      F     1   
27724  hemapenpal2016@gmail.com        lavanya         R      F     1   
27725  hemapenpal2016@gmail.com     madhimalar         B      F     1   
27726  hemapenpal2016@gmail.com        nathiya         V      F     1   
27727  hemapenpal2016@gmail.com       sandhiya         D      F     1   
27728  hemapenpal2016@gmail.com       thamarai         S      F     1   
28000  hemapenpal2016@g

<p> Things are different here. Even though they share the same email, they have more likely address and names.

### gender

In [119]:
userComb.gender.fillna(0,inplace=True)
userComb.gender=userComb.gender.map({'M':1,'F':2,0:0})
userComb.groupby('gender')['id','username'].count()

id  username
gender                 
0       34022     34022
1       44522     44522
2       48932     48932

<p>34022 people didn't input their gender.</p>
<p>0: NaN, 1: Male, 2: Famale </p>

### tier

In [171]:
userComb.groupby('tier')['id','username'].count()

id  username
tier                  
1     116264    116264
2      11207     11207
4          5         5

In [172]:
userComb[userComb.tier==4]

last_login is_superuser      id      username  \
0       2017-01-27 17:42:31.857178+00            t       1    migueldv90   
4       2017-01-07 15:51:51.854495+00            f       5     joetroyen   
24      2017-01-26 20:58:10.536884+00            t      32  markdanforth   
19808   2017-01-09 22:20:54.570048+00            f   21949      CrystalG   
100930  2016-12-16 01:03:02.073771+00            f  104036    josephyu28   

                              email first_name last_name gender  tier  \
0          miguel@penpalschools.com     Miguel   Vazquez      M     4   
4             joe@penpalschools.com        Joe    Troyen    NaN     4   
24      mark.danforth@bookaround.co     PenPal     Coach    NaN     4   
19808      crystalgermond@gmail.com    Crystal       NaN      F     4   
100930     joseph@penpalschools.com     Joseph        Yu      M     4   

                                                      img  \
0       user_app/1/08_07_2015_18_26_46_736474_upload-i...   
4       user_app/5/04_01_2015_17_59_50_754213_upload-i...   
24                                    default-profile.svg   
19808   user_app/21949/09_10_2015_18_53_10_968546_uplo...   
100930                                default-profile.svg   

                 ...           email_verified points  \
0                ...                        f      0   
4                ...                        f      0   
24               ...                        t      0   
19808            ...                        f      0   
100930           ...                        t      0   

                         img_activity street_line1 street_line2 zipcode  \
0       2016-09-09 04:17:40.075083+00          NaN          NaN     NaN   
4       2016-09-09 04:17:40.075083+00          NaN          NaN     NaN   
24             2016-09-09 04:17:40+00          NaN          NaN     NaN   
19808   2016-09-09 04:17:40.075083+00          NaN          NaN     NaN   
100930         2016-09-30 13:45:20+00          NaN          NaN     NaN   

                 city  state         country       formatted_address  
0              Austin     TX   United States         Austin, TX, USA  
4              Austin    TX   United States         Austin, TX, USA   
24             Austin     TX   United States         Austin, TX, USA  
19808          Austin     TX   United States         Austin, TX, USA  
100930  State College     PA   United States  State College, PA, USA  

[5 rows x 26 columns]

In [173]:
userComb=userComb[userComb.tier!=4]

In [174]:
userComb.describe(include=['O'])

last_login is_superuser username  \
count                         127471       127471   127471   
unique                        127465            1   127471   
top     2015-10-13 15:12:18.20282+00            f  KateNic   
freq                               2       127471        1   

                        email first_name last_name gender  \
count                   52640      97649     95662  93451   
unique                  50728      27800      3444      2   
top     porurpenpal@gmail.com      Emily         M      F   
freq                       81        457      7363  48931   

                        img      bio           stripe_id        ...         \
count                127467    15646                  57        ...          
unique                23026    15150                  57        ...          
top     default-profile.svg  Teacher  cus_9rpzLyBlQNoNUA        ...          
freq                  75449       73                   1        ...          

                                  dob email_verified  \
count                          127471         127471   
unique                          95893              2   
top     2015-03-20 03:20:34.472988+00              f   
freq                             9243         122409   

                         img_activity street_line1 street_line2 zipcode  \
count                          127471          585            5    3971   
unique                          39787          433            1    2355   
top     2016-09-09 04:17:40.075083+00     New York                48118   
freq                            87649           21            5      62   

          city  state country formatted_address  
count    24925  25632  127463             27107  
unique    6691   1846     308             10189  
top     Austin     TX    None   Austin, TX, USA  
freq       478   1192  100363               344  

[4 rows x 23 columns]

# colums:
'last_login', 'is_superuser', 'id', 'username', 'email',
       'first_name', 'last_name', 'gender', 'tier', 'img', 'bio',
       'stripe_id', 'is_staff', 'is_active', 'date_joined', 'dob',
       'email_verified', 'points', 'img_activity', 'street_line1',
       'street_line2', 'zipcode', 'city', 'state', 'country',
       'formatted_address'

In [177]:
userComb.groupby('stripe_id')['id','username'].count()

id  username
stripe_id                        
 cus_9jV0r7YqMmuujn   1         1
 cus_9jV23kA6jm9mda   1         1
cus_91YCgccMZOhNbZ    1         1
cus_91rKWextvIl3MS    1         1
cus_94dV9PkTgoSSHg    1         1
cus_95EeeYa1MS3fVS    1         1
cus_96uIk1He41SP2P    1         1
cus_99XbW3VSpUbO8i    1         1
cus_99kko8CA0ApOGJ    1         1
cus_99l90mvIpnEo4O    1         1
cus_9AiVKpCju61EeE    1         1
cus_9C8zvFgwHFRiOa    1         1
cus_9Du0MrUlhqvzDa    1         1
cus_9EgZ7KknUDjcmC    1         1
cus_9ErgjNkspJQyvc    1         1
cus_9F7l4cWPU7fu6e    1         1
cus_9HR1oyIqQJE7yv    1         1
cus_9Hg4QEl0HHZJjl    1         1
cus_9OXVO5i3wiLHGo    1         1
cus_9OZ9UWXQAj29rV    1         1
cus_9Rvj9dFVh7UnMk    1         1
cus_9StV44YLgKa2Eo    1         1
cus_9TusvVVdGD7ehW    1         1
cus_9UNnXZrFimYSC5    1         1
cus_9VvmqyoBioS7l5    1         1
cus_9XNEnUdZQqpUz5    1         1
cus_9XajlyS3fdxVLM    1         1
cus_9YhtzqIOd007Qb    1         1
cus_9fJTdwaNSbPuSB    1         1
cus_9g1f2nMdaLYhMU    1         1
cus_9gjyWUd27o8plL    1         1
cus_9hvzaMslPduJfJ    1         1
cus_9iFHSjShYHpols    1         1
cus_9jKSTzdjLEfx3m    1         1
cus_9jUzW1C20m426V    1         1
cus_9jV0UUCeSFMWpm    1         1
cus_9jV15R7iJxCBGV    1         1
cus_9jw5ZIoAh9t12j    1         1
cus_9mnVeI5K3RYd5R    1         1
cus_9osxuDHWcGk36t    1         1
cus_9qrICfVEHuykFo    1         1
cus_9rLJgUGQ1419RI    1         1
cus_9rd3HOYlUWs1iC    1         1
cus_9rpzLyBlQNoNUA    1         1
cus_9sPTbmAHtiOwn4    1         1
cus_9tceOgUvJblemh    1         1
cus_9ttMrkvyCQv63M    1         1
cus_9uKZCCPagu2wyA    1         1
cus_9ugxDTJ4equNuc    1         1
cus_9vAp1hQo7gRrdi    1         1
cus_9vP23nMUPXAvyX    1         1
cus_9vQ7UHfUes3Cu9    1         1
cus_9wJCn8zPNUbEr9    1         1
cus_9wYcG1E04n0yLY    1         1
cus_9wvkJKmNkbIJkB    1         1
cus_9wvkn6yAx8U2B8    1         1
cus_9xHf9rUR9iL7zq    1         1

Each of 57 user has their own unique strip id. So I can delete this colunm and add a new colunm to show if a user has strip.

In [ ]:
userComb=

The useful files are:
<p>
"data/course_classroom.csv","data/course_match.csv","data/course_match_student_match.csv",
"data/course_questionresponse.csv","data/course_questionresponse_new_response.csv",
"data/course_questionresponse_new_subresponse.csv","data/course_subquestionresponse.csv",
"data/user_app_address.csv","data/user_app_customuser.csv","data/user_messages_message.csv".
</p>

In [6]:
files=["data/course_match.csv","data/course_questionresponse.csv","data/course_questionresponse_new_response.csv","data/course_questionresponse_new_subresponse.csv","data/course_subquestionresponse.csv","data/user_app_address.csv","data/user_app_customuser.csv","data/user_messages_message.csv"]

In [7]:
def printFile(filePath):
    with open(filePath,'r') as f:
        #data = pd.read_csv(f,delim_whitespace=True,header=None,error_bad_lines=False)
        data = pd.read_csv(f)
        print "=================",filePath,"================="
        #print "summary:\n",data.describe(),"\n\nhead:"
        print data.head(10),'\n'
        #return data

In [8]:
for fi in files:
    printFile(fi)

================= data/course_match.csv =================
      id  student_id  classroom_id  \
0  97395      123884        8253.0   
1  99041      125418        7989.0   
2  12327       17867        1998.0   
3  12632       18857        1911.0   
4  45091       60318        4072.0   
5     50         154          69.0   
6     44         148          69.0   
7     32         134          69.0   
8     34         138          69.0   
9  12180       18321        1965.0   

                                               intro new_match  age  \
0                Australia beacause they have coalas         f   13   
1       Paris, France because it's the city of love.         f   13   
2  We are a free country.\nWe drink water.\nWe bo...         t   13   
3                                                NaN         t   15   
4                                Watch a hockey game         t   10   
5                                                NaN         f   12   
6                         

/Users/Wenbo/anaconda/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2821: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  if self.run_code(code, result):


================= data/user_app_address.csv =================
   id street_line1 street_line2 zipcode       city     state          country  \
0   2          NaN          NaN     NaN         SF       CA    United States    
1   3          NaN          NaN     NaN    Watford  England   United Kingdom    
2   4          NaN          NaN     NaN     Austin       TX    United States    
3   5          NaN          NaN     NaN     Austin       TX    United States    
4   6          NaN          NaN     NaN         CF       NaN       Argentina    
5   7          NaN          NaN     NaN        NaN       NaN             None   
6   8          NaN          NaN     NaN        NaN       NaN             None   
7   9          NaN          NaN     NaN        NaN       NaN             None   
8  39          NaN          NaN     NaN  Wadsworth        IL    United States   
9  11          NaN          NaN     NaN        NaN       NaN             None   

             formatted_address        lat     

/Users/Wenbo/anaconda/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2821: DtypeWarning: Columns (13) have mixed types. Specify dtype option on import or set low_memory=False.
  if self.run_code(code, result):


#### First, how to define active and inactive stuent?
<p>I would like to use messages tables and response tables. The word count ,the frequency, the amount of total response, etc data will be used to identify whether a student is activie or not.</p>
#### Second, what I should use for classification?
<p>their profile data. Because Joe said they want to classify the student before the matching/class </p>

### data/course_match.csv
Select most of columns and see if there is a trends.

In [2]:
data_1=pd.read_csv("data/course_match.csv",usecols=['student_id','intro','new_match','age','course_start_date','target_start_date','skill','state','country','activity','points','pledge'])

#### ? Don't know pledge. Also the id in original csv file.

In [3]:
data_1.head()

student_id                                              intro new_match  \
0      123884                Australia beacause they have coalas         f   
1      125418       Paris, France because it's the city of love.         f   
2       17867  We are a free country.\nWe drink water.\nWe bo...         t   
3       18857                                                NaN         t   
4       60318                                Watch a hockey game         t   

   age course_start_date target_start_date         skill state        country  \
0   13        2017-01-09        2017-01-09  Intermediate   NaN        Denmark   
1   13        2017-01-09        2017-01-09  Intermediate    FL  United States   
2   13        2015-05-25        2015-08-31      Beginner    AL  United States   
3   15               NaN        2015-08-31      Beginner    TX  United States   
4   10        2016-02-01        2016-01-18      Beginner    OR  United States   

                        activity  points pledge  
0  2017-01-26 09:03:08.858457+00     300      t  
1  2017-01-25 19:19:34.060776+00     300      t  
2  2015-08-30 19:15:09.928069+00     150      f  
3  2015-08-30 19:14:26.770596+00       0      f  
4   2016-01-29 20:15:01.09436+00     100      f

In [4]:
data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91259 entries, 0 to 91258
Data columns (total 12 columns):
student_id           91259 non-null int64
intro                74651 non-null object
new_match            91259 non-null object
age                  91259 non-null int64
course_start_date    76007 non-null object
target_start_date    91259 non-null object
skill                91259 non-null object
state                89681 non-null object
country              91248 non-null object
activity             91259 non-null object
points               91259 non-null int64
pledge               91259 non-null object
dtypes: int64(3), object(9)
memory usage: 8.4+ MB


In [5]:
data_1.describe()

student_id           age        points
count   91259.000000  91259.000000  91259.000000
mean    71835.266878     12.593476    203.267075
std     36299.815653     26.125831    127.896151
min        88.000000    -71.000000      0.000000
25%     38751.500000     11.000000    100.000000
50%     75364.000000     12.000000    200.000000
75%    102613.500000     14.000000    300.000000
max    130738.000000   1996.000000    450.000000

#### Hypothesis
There is 91259 records, 74651 of them have introduction, 76007 have course_start_date, 89681 have state. SO:
<li>Not all records has "state" maybe because states are not available for some foreign contry.</li>
<li>? need to know why there are some records with no course_start_date.</li>
<li> some students may have no introduction.</li>
And from the statistic data:
<li> Age data has some problems: the min value should not be negative; the max value should not be 1996, which could be born year.</li>

In [ ]:
data_1.pop()

In [9]:
data_2 = pd.read_csv("data/course_questionresponse.csv")
data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 445977 entries, 0 to 445976
Data columns (total 7 columns):
id             445977 non-null int64
created        445977 non-null object
modified       445977 non-null object
match_id       445977 non-null int64
question_id    255484 non-null float64
text           445973 non-null object
unit_id        255484 non-null float64
dtypes: float64(2), int64(2), object(3)
memory usage: 23.8+ MB


In [10]:
data_2.describe()

id       match_id    question_id        unit_id
count  445977.000000  445977.000000  255484.000000  255484.000000
mean   293172.854874   50610.124193     415.205876      32.390005
std    157232.444675   28513.655944     258.006175      21.902441
min         2.000000      32.000000       2.000000       1.000000
25%    163848.000000   25333.000000     131.000000       4.000000
50%    300091.000000   49898.000000     393.000000      40.000000
75%    428269.000000   75802.000000     701.000000      43.000000
max    556424.000000  104177.000000     732.000000      96.000000

#### Question
question response only has match_id, what does id means? How can I know the response is made by whom? So I must use match table to find match id and student id? 